In [1]:
import pandas as pd

bp_details = pd.read_csv('bp_details.csv')
all_bp = pd.read_csv('all_bp.csv')
all_oscars = pd.read_csv('all_oscars.csv')
bp_nominees = pd.read_csv('wiki_clean.csv')

In [2]:
#take a look
bp_details.head()
#drop extra index
bp_details= bp_details.drop(columns={'Unnamed: 0'})
bp_details.head()

title  runtime rating                      genre  \
0                    Nomadland  107 min      R                      Drama   
1                   The Father   97 min  PG-13             Drama, Mystery   
2  Judas and the Black Messiah  126 min      R  Biography, Drama, History   
3                         Mank  131 min      R   Biography, Comedy, Drama   
4                       Minari  115 min  PG-13                      Drama   

                                             summary gross  year  
0  A woman in her sixties, after losing everythin...   NaN  2020  
1  A man refuses all assistance from his daughter...   NaN  2020  
2  Offered a plea deal by the FBI, William O'Neal...   NaN  2021  
3  1930s Hollywood is re-evaluated through the ey...   NaN  2020  
4   A Korean family starts a farm in 1980s Arkansas.   NaN  2020

In [3]:
#all_bp = all_bp.drop(columns={'Unnamed: 0'})
all_bp = all_bp.query('year >= 1963')
all_bp = all_bp.query('year <= 2020')
all_bp = all_bp.drop(columns = {'Unnamed: 0'})
all_bp

category                           nominee  year  winner
0    Best Picture                Lawrence of Arabia  1963     1.0
1    Best Picture                   The Longest Day  1963     0.0
2    Best Picture  Meredith Willson's The Music Man  1963     0.0
3    Best Picture              Mutiny on the Bounty  1963     0.0
4    Best Picture             To Kill a Mockingbird  1963     0.0
..            ...                               ...   ...     ...
328  Best Picture                             Joker  2020     0.0
329  Best Picture                      Little Women  2020     0.0
330  Best Picture                    Marriage Story  2020     0.0
331  Best Picture                              1917  2020     0.0
332  Best Picture   Once upon a Time...in Hollywood  2020     0.0

[333 rows x 4 columns]

In [4]:
all_oscars = all_oscars.drop(columns={'Unnamed: 0'})
all_oscars.head()

category              nominee  year  winner
0    Actor        Emil Jannings  1929     1.0
1    Actor  Richard Barthelmess  1929     0.0
2  Actress         Janet Gaynor  1929     1.0
3  Actress       Louise Dresser  1929     0.0
4  Actress       Gloria Swanson  1929     0.0

In [5]:
bp_nominees = bp_nominees.drop(columns={'Unnamed: 0'})
bp_nominees.head()

film              producer  release_year
0                Wings  Famous Players-Lasky          1927
1           7th Heaven                   Fox          1927
2           The Racket     The Caddo Company          1927
3  The Broadway Melody   Metro-Goldwyn-Mayer          1928
4                Alibi   Feature Productions          1928

In [6]:
bp_nominees = bp_nominees.query('release_year >= 1962')
bp_nominees = bp_nominees.query('release_year <= 2019')
bp_nominees.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 333 entries, 230 to 562
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   film          333 non-null    object
 1   producer      333 non-null    object
 2   release_year  333 non-null    int64 
dtypes: int64(1), object(2)
memory usage: 10.4+ KB


### box office numbers and gross aren't available yet for 2020 (and is such an anaomoly, can it be useful?) Best Picture became a category in 1963, so narrow in on years 1963-2020

In [7]:
bp_details = bp_details.query('year >= 1962')
bp_details = bp_details.query('year <= 2019')
bp_details

title  runtime     rating                        genre  \
6           Sound of Metal  120 min          R                 Drama, Music   
8                 Parasite  132 min          R      Comedy, Drama, Thriller   
9           Ford v Ferrari  152 min      PG-13     Action, Biography, Drama   
10                   Joker  122 min          R       Crime, Drama, Thriller   
11          Marriage Story  137 min          R               Drama, Romance   
..                     ...      ...        ...                          ...   
335     Lawrence of Arabia  218 min   Approved  Adventure, Biography, Drama   
336        The Longest Day  178 min          G       Action, Drama, History   
337          The Music Man  151 min          G      Comedy, Family, Musical   
338   Mutiny on the Bounty  178 min  Not Rated    Adventure, Drama, History   
339  To Kill a Mockingbird  129 min   Approved                 Crime, Drama   

                                               summary        gross  year  
6    A heavy-metal drummer's life is thrown into fr...          NaN  2019  
8    Greed and class discrimination threaten the ne...   53,367,844  2019  
9    American car designer Carroll Shelby and drive...  117,624,028  2019  
10   A mentally troubled stand-up comedian embarks ...  335,451,311  2019  
11   Noah Baumbach's incisive and compassionate loo...    2,000,000  2019  
..                                                 ...          ...   ...  
335  The story of T.E. Lawrence, the English office...   44,824,144  1962  
336  The events of D-Day, told on a grand scale fro...   39,100,000  1962  
337  Traveling con artist Harold Hill targets the n...   14,953,846  1962  
338  In 1787, British ship Bounty leaves Portsmouth...   13,680,000  1962  
339  Atticus Finch, a widowed lawyer in Depression-...          NaN  1962  

[332 rows x 7 columns]

In [8]:
all_bp = all_bp.query('year >= 1963')
all_bp = all_bp.query('year <= 2021')
all_bp = all_bp.reset_index(drop=True)
all_bp

category                           nominee  year  winner
0    Best Picture                Lawrence of Arabia  1963     1.0
1    Best Picture                   The Longest Day  1963     0.0
2    Best Picture  Meredith Willson's The Music Man  1963     0.0
3    Best Picture              Mutiny on the Bounty  1963     0.0
4    Best Picture             To Kill a Mockingbird  1963     0.0
..            ...                               ...   ...     ...
328  Best Picture                             Joker  2020     0.0
329  Best Picture                      Little Women  2020     0.0
330  Best Picture                    Marriage Story  2020     0.0
331  Best Picture                              1917  2020     0.0
332  Best Picture   Once upon a Time...in Hollywood  2020     0.0

[333 rows x 4 columns]

test = pd.merge(bp_details, all_bp, left_on='title', right_on='nominee', how='outer')
test.head(60)

test[test.title.isna()]

#test merge shows some titles are different in each database - 368 in outer merge vs 303 in inner merge - 338 in bp_details vs 333 in bp_nominees
#Music Man vs Meredith Wilson's Music Man
#Who's Afraid of Virginia Woolf (?)
#MASH no special *
#Cries and (&) Whispers
#1917 gone from bp_details (reg ex error)
#Precious
#birdman
#127 hours (regex error)
#Once (...) upon a time in Hollywood

bp_details = bp_details.sort_values(['year', 'title']).reset_index(drop=True)
bp_details

all_bp = all_bp.sort_values(['year', 'nominee']).reset_index(drop=True)
all_bp

pd.concat([bp_details, all_bp], axis=1)

test2 = pd.merge(bp_details, all_bp, left_on='title', right_on='nominee', how='inner')
test2 = test2.rename(columns={'year_x':'premiere_year', 'year_y':'oscar_year'})
test2

!conda install -c conda-forge fuzzywuzzy

In [9]:
import difflib
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

def fuzzy_merge(df_1, df_2, key1, key2, threshold=90, limit=2):

#df_1 is the left table to join
#df_2 is the right table to join
#key1 is the key column of the left table
#key2 is the key column of the right table
#threshold is how close the matches should be to return a match, based on Levenshtein distance
#limit is the amount of matches that will get returned, these are sorted high to low

    s = df_2[key2].tolist()

    m = df_1[key1].apply(lambda x: process.extract(x, s, limit=limit))    
    df_1['matches'] = m

    m2 = df_1['matches'].apply(lambda x: ', '.join([i[0] for i in x if i[1] >= threshold]))
    df_1['matches'] = m2

    return df_1

In [10]:
big_df = fuzzy_merge(bp_details, all_bp, 'title', 'nominee', threshold=90, limit=1)
big_df.head()

title  runtime rating                     genre  \
6   Sound of Metal  120 min      R              Drama, Music   
8         Parasite  132 min      R   Comedy, Drama, Thriller   
9   Ford v Ferrari  152 min  PG-13  Action, Biography, Drama   
10           Joker  122 min      R    Crime, Drama, Thriller   
11  Marriage Story  137 min      R            Drama, Romance   

                                              summary        gross  year  \
6   A heavy-metal drummer's life is thrown into fr...          NaN  2019   
8   Greed and class discrimination threaten the ne...   53,367,844  2019   
9   American car designer Carroll Shelby and drive...  117,624,028  2019   
10  A mentally troubled stand-up comedian embarks ...  335,451,311  2019   
11  Noah Baumbach's incisive and compassionate loo...    2,000,000  2019   

           matches  
6                   
8         Parasite  
9   Ford v Ferrari  
10           Joker  
11  Marriage Story

In [11]:
everything = pd.merge(big_df, all_bp, left_on='matches', right_on='nominee', how='outer')

In [12]:
everything.tail(10)

title  runtime     rating                        genre  \
327     Lawrence of Arabia  218 min   Approved  Adventure, Biography, Drama   
328        The Longest Day  178 min          G       Action, Drama, History   
329          The Music Man  151 min          G      Comedy, Family, Musical   
330   Mutiny on the Bounty  178 min  Not Rated    Adventure, Drama, History   
331  To Kill a Mockingbird  129 min   Approved                 Crime, Drama   
332                    NaN      NaN        NaN                          NaN   
333                    NaN      NaN        NaN                          NaN   
334                    NaN      NaN        NaN                          NaN   
335                    NaN      NaN        NaN                          NaN   
336                    NaN      NaN        NaN                          NaN   

                                               summary       gross  year_x  \
327  The story of T.E. Lawrence, the English office...  44,824,144  1962.0   
328  The events of D-Day, told on a grand scale fro...  39,100,000  1962.0   
329  Traveling con artist Harold Hill targets the n...  14,953,846  1962.0   
330  In 1787, British ship Bounty leaves Portsmouth...  13,680,000  1962.0   
331  Atticus Finch, a widowed lawyer in Depression-...         NaN  1962.0   
332                                                NaN         NaN     NaN   
333                                                NaN         NaN     NaN   
334                                                NaN         NaN     NaN   
335                                                NaN         NaN     NaN   
336                                                NaN         NaN     NaN   

                              matches      category  \
327                Lawrence of Arabia  Best Picture   
328                   The Longest Day  Best Picture   
329  Meredith Willson's The Music Man  Best Picture   
330              Mutiny on the Bounty  Best Picture   
331             To Kill a Mockingbird  Best Picture   
332                               NaN  Best Picture   
333                               NaN  Best Picture   
334                               NaN  Best Picture   
335                               NaN  Best Picture   
336                               NaN  Best Picture   

                              nominee  year_y  winner  
327                Lawrence of Arabia  1963.0     1.0  
328                   The Longest Day  1963.0     0.0  
329  Meredith Willson's The Music Man  1963.0     0.0  
330              Mutiny on the Bounty  1963.0     0.0  
331             To Kill a Mockingbird  1963.0     0.0  
332                           M*A*S*H  1971.0     0.0  
333           The Godfather, Part III  1991.0     0.0  
334                     Winter's Bone  2011.0     0.0  
335                    Phantom Thread  2018.0     0.0  
336                              1917  2020.0     0.0

In [13]:
everything.to_csv('bp_full_df1.csv', index=False)

In [14]:
#cleared up remaining NaN's manually in Excel, saved as bp-full-df2
#next, merge wikipedia data on Producer(s) with bp_full_df2

In [15]:
bp_nominees=bp_nominees.query('release_year <= 2020')
bp_nominees = bp_nominees.query('release_year >= 1963')
bp_nominees['release_year'].min()

1963

In [16]:
bp_full_df2 = pd.read_csv('bp_full_df2.csv')
bp_full_df2 = bp_full_df2.drop(columns={'Unnamed: 0'})
bp_full_df2.head()

title  runtime rating                    genre  \
0     Jojo Rabbit  108 min  PG-13       Comedy, Drama, War   
1  Sound of Metal  120 min      R             Drama, Music   
2  Marriage Story  137 min      R           Drama, Romance   
3    The Irishman  209 min      R  Biography, Crime, Drama   
4        Parasite  132 min      R  Comedy, Drama, Thriller   

                                             summary          gross  \
0  A young German boy in the Hitler Youth whose h...     349,555.00   
1  A heavy-metal drummer's life is thrown into fr...     516,520.00   
2  Noah Baumbach's incisive and compassionate loo...   2,000,000.00   
3  Hitman Frank Sheeran looks back at the secrets...   7,000,000.00   
4  Greed and class discrimination threaten the ne...  53,367,844.00   

   release_year         matches      category         nominee  oscar_year  \
0          2019     Jojo Rabbit  Best Picture     Jojo Rabbit        2020   
1          2019  Sound of Metal  Best Picture  Sound of Metal        2020   
2          2019  Marriage Story  Best Picture  Marriage Story        2020   
3          2019    The Irishman  Best Picture    The Irishman        2020   
4          2019        Parasite  Best Picture        Parasite        2020   

   winner  
0       0  
1       0  
2       0  
3       0  
4       1

In [17]:
bp_nominees.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 328 entries, 235 to 562
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   film          328 non-null    object
 1   producer      328 non-null    object
 2   release_year  328 non-null    int64 
dtypes: int64(1), object(2)
memory usage: 10.2+ KB


In [18]:
bp_nominees.head()

film         producer  release_year
235             Tom Jones  Tony Richardson          1963
236       America America       Elia Kazan          1963
237             Cleopatra    Walter Wanger          1963
238  How the West Was Won    Bernard Smith          1963
239   Lilies of the Field     Ralph Nelson          1963

In [19]:
producers = pd.merge(bp_full_df2, bp_nominees, left_on='title', right_on='film', how='outer')
producers.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 344 entries, 0 to 343
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           335 non-null    object 
 1   runtime         335 non-null    object 
 2   rating          335 non-null    object 
 3   genre           335 non-null    object 
 4   summary         335 non-null    object 
 5   gross           326 non-null    object 
 6   release_year_x  335 non-null    float64
 7   matches         335 non-null    object 
 8   category        335 non-null    object 
 9   nominee         335 non-null    object 
 10  oscar_year      335 non-null    float64
 11  winner          335 non-null    float64
 12  film            328 non-null    object 
 13  producer        328 non-null    object 
 14  release_year_y  328 non-null    float64
dtypes: float64(4), object(11)
memory usage: 43.0+ KB


In [20]:
#335 is the right number of best picture nominees
print(335-328)
producers

7


title  runtime rating                    genre  \
0       Jojo Rabbit  108 min  PG-13       Comedy, Drama, War   
1    Sound of Metal  120 min      R             Drama, Music   
2    Marriage Story  137 min      R           Drama, Romance   
3      The Irishman  209 min      R  Biography, Crime, Drama   
4          Parasite  132 min      R  Comedy, Drama, Thriller   
..              ...      ...    ...                      ...   
339             NaN      NaN    NaN                      NaN   
340             NaN      NaN    NaN                      NaN   
341             NaN      NaN    NaN                      NaN   
342             NaN      NaN    NaN                      NaN   
343             NaN      NaN    NaN                      NaN   

                                               summary          gross  \
0    A young German boy in the Hitler Youth whose h...     349,555.00   
1    A heavy-metal drummer's life is thrown into fr...     516,520.00   
2    Noah Baumbach's incisive and compassionate loo...   2,000,000.00   
3    Hitman Frank Sheeran looks back at the secrets...   7,000,000.00   
4    Greed and class discrimination threaten the ne...  53,367,844.00   
..                                                 ...            ...   
339                                                NaN            NaN   
340                                                NaN            NaN   
341                                                NaN            NaN   
342                                                NaN            NaN   
343                                                NaN            NaN   

     release_year_x         matches      category         nominee  oscar_year  \
0            2019.0     Jojo Rabbit  Best Picture     Jojo Rabbit      2020.0   
1            2019.0  Sound of Metal  Best Picture  Sound of Metal      2020.0   
2            2019.0  Marriage Story  Best Picture  Marriage Story      2020.0   
3            2019.0    The Irishman  Best Picture    The Irishman      2020.0   
4            2019.0        Parasite  Best Picture        Parasite      2020.0   
..              ...             ...           ...             ...         ...   
339             NaN             NaN           NaN             NaN         NaN   
340             NaN             NaN           NaN             NaN         NaN   
341             NaN             NaN           NaN             NaN         NaN   
342             NaN             NaN           NaN             NaN         NaN   
343             NaN             NaN           NaN             NaN         NaN   

     winner                                             film  \
0       0.0                                      Jojo Rabbit   
1       0.0                                              NaN   
2       0.0                                   Marriage Story   
3       0.0                                     The Irishman   
4       1.0                                         Parasite   
..      ...                                              ...   
339     NaN                         The Postman (Il Postino)   
340     NaN                        Good Night, and Good Luck   
341     NaN  Precious: Based on the Novel 'Push' by Sapphire   
342     NaN  Birdman or (The Unexpected Virtue of Ignorance)   
343     NaN                    Once Upon a Time in Hollywood   

                                              producer  release_year_y  
0    Carthew Neal, Taika Waititi, and Chelsea Winst...          2019.0  
1                                                  NaN             NaN  
2                       Noah Baumbach and David Heyman          2019.0  
3    Martin Scorsese, Robert De Niro, Jane Rosentha...          2019.0  
4                         Kwak Sin-ae and Bong Joon-ho          2019.0  
..                                                 ...             ...  
339  Mario Cecchi Gori, Vittorio Cecchi Gori, and G...          1995.0  
340                                       

In [21]:
#title differences in 7 films
producers.tail(7)

title runtime rating genre summary gross  release_year_x matches category  \
337   NaN     NaN    NaN   NaN     NaN   NaN             NaN     NaN      NaN   
338   NaN     NaN    NaN   NaN     NaN   NaN             NaN     NaN      NaN   
339   NaN     NaN    NaN   NaN     NaN   NaN             NaN     NaN      NaN   
340   NaN     NaN    NaN   NaN     NaN   NaN             NaN     NaN      NaN   
341   NaN     NaN    NaN   NaN     NaN   NaN             NaN     NaN      NaN   
342   NaN     NaN    NaN   NaN     NaN   NaN             NaN     NaN      NaN   
343   NaN     NaN    NaN   NaN     NaN   NaN             NaN     NaN      NaN   

    nominee  oscar_year  winner  \
337     NaN         NaN     NaN   
338     NaN         NaN     NaN   
339     NaN         NaN     NaN   
340     NaN         NaN     NaN   
341     NaN         NaN     NaN   
342     NaN         NaN     NaN   
343     NaN         NaN     NaN   

                                                film  \
337                               Cries and Whispers   
338                          Raiders of the Lost Ark   
339                         The Postman (Il Postino)   
340                        Good Night, and Good Luck   
341  Precious: Based on the Novel 'Push' by Sapphire   
342  Birdman or (The Unexpected Virtue of Ignorance)   
343                    Once Upon a Time in Hollywood   

                                              producer  release_year_y  
337                                     Ingmar Bergman          1973.0  
338                                     Frank Marshall          1981.0  
339  Mario Cecchi Gori, Vittorio Cecchi Gori, and G...          1995.0  
340                                       Grant Heslov          2005.0  
341  Lee Daniels, Sarah Siegel-Magness, and Gary Ma...          2009.0  
342  Alejandro G. Iñárritu, John Lesher, and James ...          2014.0  
343  David Heyman, Shannon McIntosh, and Quentin Ta...          2019.0

In [22]:
#edit last 7 in excel to expedite process

In [23]:
producers.to_csv('with_producers2.csv')

In [24]:
producers = pd.read_csv('with_producers.csv')

In [25]:
producers = producers.drop(columns={'release_year_y', 'nominee', 'matches'})
producers = producers.drop(columns={'Unnamed: 0'})
producers

title  runtime     rating                        genre  \
0              Jojo Rabbit  108 min      PG-13           Comedy, Drama, War   
1           Marriage Story  137 min          R               Drama, Romance   
2             The Irishman  209 min          R      Biography, Crime, Drama   
3                 Parasite  132 min          R      Comedy, Drama, Thriller   
4             Little Women  135 min         PG               Drama, Romance   
..                     ...      ...        ...                          ...   
329   Mutiny on the Bounty  178 min  Not Rated    Adventure, Drama, History   
330          The Music Man  151 min          G      Comedy, Family, Musical   
331        The Longest Day  178 min          G       Action, Drama, History   
332     Lawrence of Arabia  218 min   Approved  Adventure, Biography, Drama   
333  To Kill a Mockingbird  129 min   Approved                 Crime, Drama   

                                               summary           gross  \
0    A young German boy in the Hitler Youth whose h...      349,555.00   
1    Noah Baumbach's incisive and compassionate loo...    2,000,000.00   
2    Hitman Frank Sheeran looks back at the secrets...    7,000,000.00   
3    Greed and class discrimination threaten the ne...   53,367,844.00   
4    Jo March reflects back and forth on her life, ...  108,101,214.00   
..                                                 ...             ...   
329  In 1787, British ship Bounty leaves Portsmouth...   13,680,000.00   
330  Traveling con artist Harold Hill targets the n...   14,953,846.00   
331  The events of D-Day, told on a grand scale fro...   39,100,000.00   
332  The story of T.E. Lawrence, the English office...   44,824,144.00   
333  Atticus Finch, a widowed lawyer in Depression-...      592,237.00   

     release_year_x      category  oscar_year  winner                   film  \
0              2019  Best Picture        2020       0            Jojo Rabbit   
1              2019  Best Picture        2020       0         Marriage Story   
2              2019  Best Picture        2020       0           The Irishman   
3              2019  Best Picture        2020       1               Parasite   
4              2019  Best Picture        2020       0           Little Women   
..              ...           ...         ...     ...                    ...   
329            1962  Best Picture        1963       0   Mutiny on the Bounty   
330            1962  Best Picture        1963       0          The Music Man   
331            1962  Best Picture        1963       0        The Longest Day   
332            1962  Best Picture        1963       1     Lawrence of Arabia   
333            1962  Best Picture        1963       0  To Kill a Mockingbird   

                                              producer  
0    Carthew Neal, Taika Waititi, and Chelsea Winst...  
1                       Noah Baumbach and David Heyman  
2    Martin Scorsese, Robert De Niro, Jane Rosentha...  
3                         Kwak Sin-ae and Bong Joon-ho  
4                                           Amy Pascal  
..                                                 ...  
329                                    Aaron Rosenberg  
330                                    Morton Da Costa  
331                                   Darryl F. Zanuck  
332                                        Sam Spiegel  
333                                     Alan J. Pakula  

[334 rows x 12 columns]

In [27]:
producers.to_csv('with_producers.csv', index=False)